### PR TITLE
[Optimization] Reduce mememory consumption in lp inference

### PR DESCRIPTION
**Pull request by isratnisa**
Resolves https://github.com/awslabs/graphstorm/issues/161 

The prediction function in link prediction inference pipeline `lp_mini_batch_predict()` saves the `scores` of all positive and negative edges which is later used in the evaluation function to calculate the `ranking`. In this PR, we directly save the `rankings` instead of `scores` which is much smaller in size. This optimization reduces memory consumption significantly. 

#### hyperparameters:
-num_negative_edges_eval: 100
-negative_sampler: joint
-eval_etype: "author,writes,paper"

##### Save `scores`:
number of testing edges = 714,566
Size of pos scores = 714,566 x 1
Size of neg scores = 714,566 x 200 
Total memory req:  (714566 + 714566 x 200) x 4 Byte = 574 MB 

##### Save `rankings` (proposed PR):
With the proposed PR, we only save rankings tensor which has a shape of [714566]. 
Total memory req:  714566 x 4 Byte = 2.85 MB

### CODE REVIEW

File: python/graphstorm/eval/evaluator.py [hunk 1]
```diff
@@ -604,7 +604,7 @@ def setup_task_tracker(self, client):
         self.tracker = client
 
     @abc.abstractmethod
-    def evaluate(self, val_scores, test_scores, total_iters):
```
Comment (classicsong) [inline python/graphstorm/eval/evaluator.py:607]: You should not change val_scores and test_scores to others. Each Evaluator has evaluate func and has the same set of arguments. This is a an abstract function.

File: python/graphstorm/eval/evaluator.py [hunk 2]
```diff
@@ -845,17 +837,15 @@ def compute_score(self, scores, train=False): # pylint:disable=unused-argument
             return_metrics[metric] = return_metric.item()
         return return_metrics
 
-    def evaluate(self, val_scores, test_scores, total_iters):
```
Comment (classicsong) [inline python/graphstorm/eval/evaluator.py:848]: Do not change the naming of val_scores and test_scores.

File: python/graphstorm/model/lp_gnn.py [hunk 3]
```diff
@@ -126,9 +127,8 @@ def lp_mini_batch_predict(model, emb, loader, device):
 
         Returns
         -------
-        dict of (list, list):
-            Return a dictionary of edge type to
-            (positive scores, negative scores)
+        rankings: Tensor
+            Rankings of positive scores
```
Comment (classicsong) [inline python/graphstorm/model/lp_gnn.py:131]: Is this a dictionary of rankings?

File: python/graphstorm/model/lp_gnn.py [hunk 4]
```diff
@@ -145,4 +145,11 @@ def lp_mini_batch_predict(model, emb, loader, device):
                     scores[canonical_etype].append(s)
                 else:
                     scores[canonical_etype] = [s]
-    return scores
+        # calculate ranking. etype is ignored in calculating ranking and
+        # global mrr. User can develop its own per etype MRR evaluator
+        rankings = []
```
Comment (classicsong) [inline python/graphstorm/model/lp_gnn.py:150]: Can we return separated ranking tensors for different etypes?
Previous users only need to modify the evaluator to develop a per etype MRR evaluator. But now they have to modify the lp_mini_batch_predict, which is more complex.

File: python/graphstorm/model/lp_gnn.py (hunk 3 above)
Comment (isratnisa) [reply python/graphstorm/model/lp_gnn.py]: Now it is.

File: python/graphstorm/eval/evaluator.py (hunk 1 above)
Comment (isratnisa) [reply python/graphstorm/eval/evaluator.py]: Reverted.

File: python/graphstorm/eval/evaluator.py (hunk 2 above)
Comment (isratnisa) [reply python/graphstorm/eval/evaluator.py]: Ok. Reverted.

File: python/graphstorm/eval/evaluator.py [hunk 5]
```diff
@@ -614,11 +614,9 @@ def evaluate(self, val_scores, test_scores, total_iters):
         Parameters
         ----------
         val_scores: dict of (list, list)
```
Comment (classicsong) [inline python/graphstorm/eval/evaluator.py:616]: please change the doc string of val_scores' data type.

File: python/graphstorm/eval/evaluator.py [hunk 6]
```diff
@@ -614,11 +614,9 @@ def evaluate(self, val_scores, test_scores, total_iters):
...
-            The positive and negative scores of validation edges
-            for each edge type
+            The rankings of validation edges for each edge type.
         test_scores: dict of (list, list)
```
Comment (classicsong) [inline python/graphstorm/eval/evaluator.py:618]: Same here.

File: python/graphstorm/eval/evaluator.py [hunk 7]
```diff
@@ -850,12 +847,10 @@ def evaluate(self, val_scores, test_scores, total_iters):
...
-        val_scores: dict of (list, list)
-            The positive and negative scores of validation edges
-            for each edge type
-        test_scores: dict of (list, list)
-            The positive and negative scores of testing edges
-            for each edge type
+        val_scores: dict of tensors
+            Rankings of positive scores of validation edges.
+        test_scores: dict of tensors
+            Rankings of positive scores of test edges.
```
Comment (classicsong) [inline python/graphstorm/eval/evaluator.py:853]:
```suggestion
            Rankings of positive scores of test edges for each edge type.
```

File: python/graphstorm/eval/evaluator.py [hunk 8]
```diff
@@ -850,12 +847,10 @@ def evaluate(self, val_scores, test_scores, total_iters):
...
         Parameters
         ----------
-        val_scores: dict of (list, list)
-            The positive and negative scores of validation edges
-            for each edge type
-        test_scores: dict of (list, list)
-            The positive and negative scores of testing edges
-            for each edge type
+        val_scores: dict of tensors
+            Rankings of positive scores of validation edges.
```
Comment (classicsong) [inline python/graphstorm/eval/evaluator.py:851]:
```suggestion
            Rankings of positive scores of validation edges for each edge type.
```

**Review by classicsong — COMMENTED** [review]
Can you put some profiling numbers in the PR description?

**Review by classicsong — APPROVED** [review]
LGTM

**Pull request merged**